src/configupdater/block.py

```diff
@@ -0,0 +1,131 @@
+
+T = TypeVar("T")
+B = TypeVar("B", bound="Block")
+
+
+class Block(ABC, Generic[T]):
+    """Abstract Block type holding lines
+
+    Block objects hold original lines from the configuration file and hold
+    a reference to a container wherein the object resides.
+
+    The type variable ``T`` is a reference for the type of the sibling blocks
+    inside the container.
+    """
+
+    def __init__(self, container: "Container[T]"):
+        self._container = container
+        self._lines: List[str] = []
+        self._updated = False
+
+    def __str__(self) -> str:
+        return "".join(self._lines)
+
+    def __repr__(self) -> str:
+        return f"<{self.__class__.__name__}: {str(self)!r}>"
+
+    def __eq__(self, other) -> bool:
+        if isinstance(other, self.__class__):
+            return str(self) == str(other)
+        else:
+            return False
+
+    def add_line(self: B, line: str) -> B:
+        """Add a line to the current block
+
+        Args:
+            line (str): one line to add
+        """
+        self._lines.append(line)
+        return self
+
+    @property
+    def lines(self) -> List[str]:
+        return self._lines
+
+    @property
+    def container(self) -> "Container[T]":
+        """Container holding the block"""
+        return self._container
+
+    @property
+    def container_idx(self: B) -> int:
+        """Index of the block within its container"""
+        return self._container.structure.index(self)
+
+    @property
+    def updated(self) -> bool:
+        """True if the option was changed/updated, otherwise False"""
+        # if no lines were added, treat it as updated since we added it
+        return self._updated or not self._lines
+
+    def _builder(self, idx: int) -> "BlockBuilder":
+        from .builder import BlockBuilder
+
+        return BlockBuilder(self._container, idx)
+
+    @property
+    def add_before(self) -> "BlockBuilder":
+        """Block builder inserting a new block before the current block"""
+        return self._builder(self.container_idx)
+
+    @property
+    def add_after(self) -> "BlockBuilder":
+        """Block builder inserting a new block after the current block"""
+        return self._builder(self.container_idx + 1)
+
+    @property
+    def next_block(self) -> Optional[T]:
+        """Next block in the current container"""
+        idx = self.container_idx + 1
+        if idx < len(self._container):
+            return self._container.structure[idx]
+        else:
+            return None
+
+    @property
+    def previous_block(self) -> Optional[T]:
+        """Previous block in the current container"""
+        idx = self.container_idx - 1
+        if idx >= 0:
+            return self._container.structure[idx]
+        else:
+            return None
+
+    def remove(self: B) -> B:
+        """Remove this block from container"""
+        self.container._remove_block(self.container_idx)
+        return self
+
+
+class Hidden(Block[T]):
```

Comment:
This class is not exactly needed. For some reason I thought it would be helpful to have a parent class in the case we need to implement common logic between the too or checking with "isinstance".
But I guess I never found that.

Do you prefer if I remove it?

Reply:
I have no strong opinion about it but would rather remove it when it has no use.

Reply:
I have removed them in the latest commits.